### DISCUSSION TITLE
DataLoaders  in Spring

### DISCUSSION THREAD

**Discussion by hantsy** [ExpediaGroup/graphql-kotlin#1533] [Q&A] [answered]
I used the Graphql Kotlin spring server in my project.

My Graphql Kotlin model is like this.

```gql
data class Post(
    val id: UUID?,
    @UpperCase
    val title: String,
    val content: String?,
    val status: String? = null,
    val createdAt: LocalDateTime?,
    val author: Author? = null,
    val authorId: UUID? = null,
    val comments: List<Comment>? = emptyList()
)

//enum class PostStatus {
//    DRAFT, PENDING_MODERATION, PUBLISHED;
//}

data class Author(
    val id: UUID?,
    val name: String,
    val email: String,
    val createdAt: LocalDateTime? = null,
    val posts: List<Post>? = emptyList()
)

data class Comment(
    val id: UUID?,
    val content: String,
    val createdAt: LocalDateTime? = null,
    val postId: UUID? = null
)

data class CreatePostInput(
    val title: String,
    val content: String,
)

data class CommentInput(
    val content: String,
    val postId: UUID
)
```

The `author` and `comments` are expected to be fetched from the lazy data loaders.

```kotlin
@Component
class AuthorsDataLoader(val authorService: AuthorService) : KotlinDataLoader<UUID, Author> {
    companion object {
        const val name = "AuthorsDataLoader"
    }

    override val dataLoaderName = name

    override fun getDataLoader(): DataLoader<UUID, Author> {
        return DataLoaderFactory.newDataLoader { keys, environment ->
            loaderScope.future {
                authorService.getAuthorByIdIn(keys).toList()
            }
        }
    }
}


@Component
class CommentsDataLoader(val postService: PostService) : KotlinDataLoader<UUID, List<Comment>> {
    companion object {
        private val log = LoggerFactory.getLogger(CommentsDataLoader::class.java)
        const val name = "CommentsDataLoader"
    }

    override val dataLoaderName = AuthorsDataLoader.name

    override fun getDataLoader(): DataLoader<UUID, List<Comment>> {
        return DataLoaderFactory.newMappedDataLoader { keys, environment ->
            loaderScope.future {
                val comments = postService.getCommentsByPostIdIn(keys).toList()
                val mappedComments: MutableMap<UUID, List<Comment>> = mutableMapOf()
                keys.forEach { mappedComments[it] = comments.filter { c -> c.postId == it } }
                log.info("mapped comments: {}", mappedComments)
                mappedComments
            }
        }
    }
}
```

I tried to  create dataFetchers for them.

```kotlin 
@Component
class CommentsDataFetcher : DataFetcher<CompletableFuture<List<Comment>>> {

    override fun get(environment: DataFetchingEnvironment): CompletableFuture<List<Comment>> {
        val postId = environment.getSource<Post>().id
        return environment.getValueFromDataLoader(CommentsDataLoader.name, postId)
    }
}

@Component
class AuthorDataFetcher : DataFetcher<CompletableFuture<Author>> {

    override fun get(environment: DataFetchingEnvironment): CompletableFuture<Author> {
        val authorId = environment.getSource<Post>().authorId
        return environment.getValueFromDataLoader(AuthorsDataLoader.name, authorId)
    }
}
```
But this does not work, in the above codes, it specified the Source(parent type) in codes.

I have explored [Gradle Kotlin Example](https://github.com/ExpediaGroup/graphql-kotlin/tree/master/examples), it created another SpringDataFetcherFactory, is it a must?

In the code first framework, such Graphql Spqr, Microprofile Graphql, it uses a `@Source` on an argument directly to archive this purpose simply. And other schema first framework, it is easy to connect the parent property with data fetchers via a RuntineWiring.

**Comment by hantsy** [comment]
After exploring the example(Spring server), I copied all files in this [executions](https://github.com/ExpediaGroup/graphql-kotlin/tree/master/examples/server/spring-server/src/main/kotlin/com/expediagroup/graphql/examples/server/spring) folder, and register a `DataFetcherFactoryProvider` bean. But it still does not work.

Is there a simple way to resolve `Post/comments` and `Post/author` via dataloaders?

**Reply by samuelAndalon** [reply]
i recommend to not use a custom `DataFetcherFactoryProvider` and work with what the schema generator does for you, and write your classes and let the data fetchers / resolvers to be autogenerated

```kotlin
@Component
class SimpleQuery : Query {
    fun comments(postId: Int, dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<List<Comment?>> =
        dataFetchingEnvironment.getValueFromDataLoader("YourDataLoaderName", postId)
}
```

**Reply by samuelAndalon** [reply]
have a look at 
https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/

and check how we recommend the usage of data loaders

**Reply by hantsy** [reply]
I want to know how to make the example style working in a Spring application.

```kotlin 
class User(val id: ID) {
    @GraphQLDescription("Get the users friends using data loader")
    fun getFriends(dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<List<User>> {
        return dataFetchingEnvironment.getValueFromDataLoader("FriendsDataLoader", id)
    }
}
```

**Reply by hantsy** [reply]
> i recommend to not use a custom `DataFetcherFactoryProvider` and work with what the schema generator does for you, and write your classes and let the data fetchers / resolvers to be autogenerated
> 
> ```kotlin
> @Component
> class SimpleQuery : Query {
>     fun comments(postId: Int, dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<List<Comment?>> =
>         dataFetchingEnvironment.getValueFromDataLoader("YourDataLoaderName", postId)
> }
> ```

I have browsed the whole doc, but not found this solution in Spring section or the example codes.

**Reply by samuelAndalon** [reply]
yeah, i just wrote that as a minimal example, if you provide a repo with basic repro steps ill be happy to take a look.

**Reply by hantsy** [reply]
I tried to do this in my exmaple: https://github.com/hantsy/spring-graphql-sample/blob/master/graphql-kotlin/src/main/kotlin/com/example/demo/gql/types/Models.kt#L23-L30

**Reply by samuelAndalon** [reply]
thats a big repository -- there you go, created this for you :) 
https://github.com/samuelAndalon/graphql-kotlin-batching-demo

**Comment by samuelAndalon** [comment]
Hello @hantsy please try to use a single discussion for the same topic, you have opened this
https://github.com/ExpediaGroup/graphql-kotlin/discussions/1533 and https://github.com/ExpediaGroup/graphql-kotlin/discussions/1534

having said that, do you have an example repo with a consistent repro ?
i cant copy paste your code bc there are some missing references how `loaderScope` is getting defined ? how `postService` looks like ?

**Comment by samuelAndalon** [comment, marked as answer]
You need very careful about the async thread model as graphql-java async model is `CompletableFuture`, if your source uses suspendable functions you cannot just interop the suspend function with a future, instead you need to run the suspend fn in a coroutine scope and use `Deferred` -- please check https://github.com/samuelAndalon/graphql-kotlin-batching-demo/blob/batching-showcase/src/main/kotlin/com/example/demo/DataLoaderConfiguration.kt#L47

that way you can get a `Deferred` Object and then you interop the Deferred with a `CompletableFuture`

TLTR; you are wrong here

```kotlin
override fun getDataLoader(): DataLoader<UUID, Author> {
        return DataLoaderFactory.newDataLoader { keys, environment ->
            loaderScope.future {
                authorService.getAuthorByIdIn(keys).toList()
            }
        }
    }
```
you need to use `Deferred` if you need to execute a suspend function and do 

```kotlin
override fun getDataLoader(): DataLoader<UUID, Author> {
        return DataLoaderFactory.newDataLoader { keys, environment ->
            loaderScope.async {
                authorService.getAuthorByIdIn(keys).toList()
            }.asCompletableFuture()
        }
    }
```

this is where suspend functions are incompatible with `CompletableFuture`, you dont need to suspend execution, you need to synchronously return a Promise and let the GraphQL engine to continue execution, having said that, i suspect your example simply hangs execution, which, makes sense given that you are suspending a function that will never execute given that the data loader registry will never dispatch

**Comment by hantsy** [comment]
I tried to add the following to my Query, 

```kotlin 
    fun comments(postId:UUID, environment: DataFetchingEnvironment): CompletableFuture<List<Comment>> {
//        val postId = environment.getSource<Post>().id
//        log.debug("Fetching comments of post: $postId")
        return environment.getValueFromDataLoader(CommentsDataLoader.name, postId)
    }

    fun author(authorId:UUID, environment: DataFetchingEnvironment): CompletableFuture<Author> {
//        val authorId = environment.getSource<Post>().authorId
//        log.debug("Fetching author of post: $authorId")
        return environment.getValueFromDataLoader(AuthorsDataLoader.name, authorId)
    }
```
In fact, I do not want these two items generated as queries, they are fields of `Post`.

```kotlin 
data class Post(
    val id: UUID?,
    @UpperCase
    val title: String,
    val content: String?,
    val status: String? = null,
    val createdAt: LocalDateTime?,
    val authorId: UUID? = null,
    val author: Author? = null,
    val comments: List<Comment>? = emptyList(),
)
```
It does not work. 

And from the coding style, I do not know where these functions are connected to the parent object `Post` at runtime.

**Reply by dariuszkuc** [reply]
> In fact, I do not want these two items generated as queries, they are fields of Post.

Move those 2 functions under `Post` class

```kotlin
class Post(...) {
  fun comments(...) { }
  fun author(...) { }
}
```

**Reply by hantsy** [reply]
Awesome, works.
